### PR TITLE
Forward compatibility with react/promise 3

### PR DESCRIPTION
**Pull request by WyriHaximus**

### CODE REVIEW

**Comment by clue** [comment]
@WyriHaximus Good initiative and would love to see this getting some traction again! What are your plans for this PR? It looks like this is currently the only PR that prevents the Socket component from also supporting Promise v3.

**Comment by WyriHaximus** [comment]
@clue Will update it tonight :+1:

**Comment by WyriHaximus** [comment]
@clue It also depends on https://github.com/reactphp/promise-timer/pull/54 to be tagged, but I've addressed the last two failures in this PR and pushed it in mergable state.

File: src/UnwrapReadableStream.php [hunk 1]
```diff
@@ -126,7 +126,7 @@ public function close()
         $this->closed = true;
 
         // try to cancel promise once the stream closes
-        if ($this->promise instanceof CancellablePromiseInterface) {
+        if ($this->promise instanceof CancellablePromiseInterface  || ($this->promise !== null && method_exists($this->promise, 'cancel') && $this->promise instanceof PromiseInterface)) {
```
Comment (clue) [inline src/UnwrapReadableStream.php:129]: Just looked into different ways to add a proper type check that works across all supported Promise versions and came up with this suggestion in https://github.com/reactphp/promise-timer/pull/55. WDYT?

```suggestion
        if ($this->promise !== null && \method_exists($this->promise, 'cancel')) {
```

This allows us to remove all references to `CancellablePromiseInterface` (see also same change below and `use` at top).

Comment (WyriHaximus) [reply src/UnwrapReadableStream.php:129]: @clue Making those changes now :+1:

File: src/UnwrapWritableStream.php [hunk 2]
```diff
@@ -153,7 +153,7 @@ public function close()
         $this->closed = true;
 
         // try to cancel promise once the stream closes
-        if ($this->promise instanceof CancellablePromiseInterface) {
+        if ($this->promise instanceof CancellablePromiseInterface  || (method_exists($this->promise, 'cancel') && $this->promise instanceof PromiseInterface)) {
```
Comment (clue) [inline src/UnwrapWritableStream.php:156]: Looks like a leftover, same as above?

Comment (WyriHaximus) [reply src/UnwrapWritableStream.php]: Yup, already made the change, but it somehow didn't get committed 0_o

Comment (clue) [reply src/UnwrapWritableStream.php]: Looks like there's still a leftover `use CancellablePromiseInterface` in this file now?

Comment (WyriHaximus) [reply src/UnwrapWritableStream.php]: Not anymore :+1:

**Review by clue — APPROVED** [review]
@WyriHaximus Thanks for the update, changes LGTM! :shipit:

**Pull request merged**